src/string_bytes.cc

```diff
@@ -568,18 +590,57 @@ static void force_ascii(const char* src, char* dst, size_t len) {
   }
 }
 
+#define packed(byte) (static_cast<uint8_t>(byte) * 0x01010101u)
+uint32_t nibbles_to_hex(uint32_t nibbles) {
+  const uint32_t ascii09 = nibbles + packed('0');
+  const uint32_t correction = packed('a' - '0' - 10);
+
+  const uint32_t tmp = nibbles + packed(128 - 10);
+  const uint32_t msb = tmp & packed(0x80);
+  const uint32_t mask = msb - (msb >> 7);
 
+  return ascii09 + (mask & correction);
+}
+#undef packed
 static size_t hex_encode(const char* src, size_t slen, char* dst, size_t dlen) {
   // We know how much we'll write, just make sure that there's space.
   CHECK(dlen >= slen * 2 &&
       "not enough space provided for hex encode");
 
   dlen = slen * 2;
-  for (uint32_t i = 0, k = 0; k < dlen; i += 1, k += 2) {
-    static const char hex[] = "0123456789abcdef";
-    uint8_t val = static_cast<uint8_t>(src[i]);
-    dst[k + 0] = hex[val >> 4];
-    dst[k + 1] = hex[val & 15];
+
+  const uint16_t* src16 = reinterpret_cast<const uint16_t*>(src);
+  size_t i = 0;
+  size_t nb = 0;
+
+  while (slen - nb >= 2) {
+#if __BYTE_ORDER == __LITTLE_ENDIAN
```

Comment:
Can you use `if (IsBigEndian()) { ... } else { ... }`?  It's a compile time expression.

(Note to self: can be constexpr now.)